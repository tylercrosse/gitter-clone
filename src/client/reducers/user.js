import { SIGN_IN } from '../actions';

const initialState = {
  loggedIn: false
};

const user = (state = initialState, action) => {
  switch (action.type) {
    case SIGN_IN:
      return {
        ...state,
        id: action.id,
        username: action.username,
        loggedIn: true
      };
    case 'SIGN_OUT':
      return {
        ...state,
        loggedIn: false
      };
    default:
      return state;
  }
};

export default user;
