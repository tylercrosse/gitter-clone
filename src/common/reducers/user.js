import { push }      from 'react-router-redux';
import { CALL_API,
  getJSON }          from 'redux-api-middleware';

// actions
export const SIGN_OUT = 'SIGN_OUT';
export const SIGN_IN_REQUEST = 'SIGN_IN_REQUEST';
export const SIGN_IN_SUCCESS = 'SIGN_IN_SUCCESS';
export const SIGN_IN_FAILURE = 'SIGN_IN_FAILURE';

// action creators
export const signOut = () => (dispatch) => {
  dispatch({ type: SIGN_OUT });
  dispatch(push('/'));
};

// FIXME change to something my expressive like 'postSignIn'
export const signIn = (username) => ({
  [CALL_API]: {
    endpoint: window.location.origin + '/api/signIn',
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ username }),
    types: [
      SIGN_IN_REQUEST,
      {
        type: SIGN_IN_SUCCESS,
        payload: /* istanbul ignore next */ (action, state, res) => getJSON(res)
      },
      SIGN_IN_FAILURE
    ]
  }
});

// reducers
const initialState = {
  loggedIn: false
};

const user = (state = initialState, action) => {
  switch (action.type) {
    case SIGN_IN_SUCCESS:
      return {
        ...state,
        ...action.payload,
        loggedIn: true
      };
    case SIGN_OUT:
      return {
        loggedIn: false
      };
    default:
      return state;
  }
};

export default user;
