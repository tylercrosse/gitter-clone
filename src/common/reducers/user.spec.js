import user from './user';

const setup = () => {
  const initialState = {
    loggedIn: false
  };

  const payload = {
    name: 'Dan',
    id: '123145'
  };

  return {
    initialState,
    payload
  };
};

describe('user reducer', () => {
  it('should handle initial state', () => {
    const { initialState } = setup();
    expect(
      user(undefined, {})
    ).toEqual(initialState);
  });

  it('should handle SIGN_IN', () => {
    const { initialState, payload } = setup();
    expect(
      user(initialState, {
        type: 'SIGN_IN',
        payload
      })
    ).toMatchSnapshot();
  });
});
