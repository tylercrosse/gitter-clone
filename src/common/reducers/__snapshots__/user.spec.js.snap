// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user reducer should handle SIGN_IN 1`] = `
Object {
  "loggedIn": false,
}
`;
